Improve cascade backrefs warning and add `code` to deprecation warnings

sirosen opened the pull request:
Fixes #6148 

The goal is to get an improved warning for users about "backref cascades" in 1.4 when `SQLALCHEMY_WARN_20=1`.

Per @zzzeek's advice in that issue (again, huge thanks for helping onboard someone new into contributing!), I've
- updated the warning text
- updated the doc on ORM cascades to mention `backref.cascade_backrefs` in addition to `relationship.cascade_backrefs`
- added mention of `backref.cascade_backrefs` to the 1.4 migration docs
- added a `code` to deprecation warnings, so that the relevant warning links to doc on ORM cascades

Most of the work is done in separate commits to make review easy, but I can rebase and squash if we intend to keep the change history.

The resulting warning now looks like so:
```
sqlalchemy.exc.RemovedIn20Warning: "Address" object is being merged into a Session along the backref cascade path for relationship "User.addresses"; in SQLAlchemy 2.0, this reverse cascade will not take place.  Set cascade_backrefs to False in either the relationship() or backref() function for the 2.0 behavior; or to set globally for the whole Session, set the future=True flag (Background on SQLAlchemy 2.0 at: http://sqlalche.me/e/b8d9) (Background on this error at: http://sqlalche.me/e/14/s9r1)
```

---

There are three parts of this that I'm not certain about, and I'd like to ask for advice/help getting this into a merge-ready state.

1. Adding `code` to warning objects

I looked at the stdlib's `warnings` to see how warnings are printed. [This line puts warnings in a small wrapper](https://github.com/python/cpython/blob/0983e01837714524fb164e784a8e96a2bc4bdf94/Lib/warnings.py#L394) and [then typically this is how that wrapper gets string-formatted](https://github.com/python/cpython/blob/0983e01837714524fb164e784a8e96a2bc4bdf94/Lib/warnings.py#L36-L37).

So all that we need to adjust is how `RemovedIn20Warning.__str__` behaves. That could be done by modifying the warning text, or by adjusting `__str__` directly. I opted for the latter, so `SADeprecationWarning` now has a modified `__str__` method.

I chose to do this by pulling `code` and `_code_str` from `SQLAlchemyError` into a mixin and then sharing that between both `SQLAlchemyError` and `SADeprecationWarning`. I wasn't sure if this would be preferable, vs copying the `code` and `_code_str` components into `SADeprecationWarning` without using the mixin, or some other approach.

I'm also not certain that this is all wanted or not?
It's simpler to just modify the message if `code != None` in `sqlalchemy.util.deprecations`, but this puts plumbing in place for deprecations to have a `code` in general.

(Aside: If the warnings have custom `__str__` methods, maybe the special-case message text for `RemovedIn20Warning` should be handled there as well?)

2. Testing

This passes existing tests today (at least, those which run under `tox -e py39-sqlite`, but it probably doesn't damage anything else).

At a minimum, some new test should cover the `SADeprecationWarning.code` behavior.

I looked at the existing tests for the ORM cascade and for deprecations, and I was a bit overwhelmed. I would like to add a test that verifies that `http://sqlalche.me/e/{version}/s9r1` appears in the deprecation warnings now, but I don't know how to do so. Any guidance on how to test this change would be a big help.

3. Are the new docs good?

The goal of these additions is primarily to make it clear that `backref.cascade_backrefs` exists, and you can read up on it if necessary. Was that accomplished? Would more narrative doc be helpful?
As a user, I've only interacted with these features very minimally, so I'm not confident that I captured everything important or stated things well.

sirosen committed:
Update warning text for cascade_backrefs=False

sirosen committed:
Document backref() for setting cascade_backrefs

Add a section to the documentation for the cascade_backrefs
parameter to `relationship()` which explains that `cascade_backrefs`
can also be used as a parameter to the `backref()` function when a
relationship is being specified using `relationship.backref`.

sirosen committed:
Expand doc on 1.4 migration change 5150

This section only describes using `relationship.cascade_backrefs`
today. Update to also mention (briefly) `backref.cascade_backrefs` for
completeness.

sirosen committed:
Add support for 'code' in deprecation warnings

SADeprecationWarning and SQLAlchemyError now share 'code' and
'_code_str' through a small mixin/helper class.

In the case of the deprecation warnings, the code_str is injected into
`__str__`, which means the message will be built when a warning is
emitted. An alternative approach would be to build a distinct message
when the warning is instantiated.

`s9r1` is a new error code which points directly to the documentation
on backref cascades, and this is now in place as the code for
`cascade_backrefs` warnings.

Fixes: #6148

zzzeek commented:
oh wow hey, thanks for doing this.

a lot to review here so I'll try a few things first.

zzzeek commented on lib/sqlalchemy/exc.py:
```diff
@@ -19,16 +19,16 @@
 
-class SQLAlchemyError(Exception):
-    """Generic error class."""
+class _CodeStrMixin(object):
```
I would call this HasDescriptionCode, and we likely don't need to extend "object" anymore, as it's a mixin so Py2 isn't going to care.

zzzeek commented on lib/sqlalchemy/orm/unitofwork.py:
```diff
@@ -26,8 +26,10 @@ def _warn_for_cascade_backrefs(state, prop):
         "reverse cascade will not take place.  Set cascade_backrefs to "
-        "False for the 2.0 behavior; or to set globally for the whole "
-        "Session, set the future=True flag" % (state.class_.__name__, prop)
+        "False in either the relationship() or backref() function for "
```
OK great.  these two functions can be linked like this:

    in either the :func:`_orm.relationship` or 
    :func:`_orm.backref` functions

zzzeek commented on lib/sqlalchemy/orm/unitofwork.py:
```diff
@@ -26,8 +26,10 @@ def _warn_for_cascade_backrefs(state, prop):
-        "False for the 2.0 behavior; or to set globally for the whole "
-        "Session, set the future=True flag" % (state.class_.__name__, prop)
+        "False in either the relationship() or backref() function for "
+        "the 2.0 behavior; or to set globally for the whole "
```
"the 2.0 behavior.   To set globally for the whole 
    :class:`_orm.Session`, set the :paramref:`_orm.Session.future` 
    parameter to ``True``.

zzzeek commented on doc/build/orm/cascades.rst:
```diff
@@ -603,7 +605,18 @@ option may be helpful for situations where an object needs to be kept out of a
 session until it's construction is completed, but still needs to be given
 associations to objects which are already persistent in the target session.
 
+When relationships are created by the ``backref`` parameter to
```
When relationships are created by the :paramref:`_orm.relationship.backref` 
    parameter on :func:`_orm.relationship`, the :paramref:`_orm.cascade_backrefs`
    parameter may be set to ``False`` on the backref side by using the 
    :func:`_orm.backref` function instead of a string.  For example, the above 
    relationship could be declared::

    .. example ..

zzzeek reviewed (commented):
comments on the code, let me look at your comment now and see what's there

zzzeek commented:
> 
> Most of the work is done in separate commits to make review easy, but I can rebase and squash if we intend to keep the change history.

we pull these PRs into our gerrit review system which does all the squashing.  you can put code into this PR in whatever way you'd like.


> 
> The resulting warning now looks like so:
> 
> ```
> sqlalchemy.exc.RemovedIn20Warning: "Address" object is being merged into a Session along the backref cascade path for relationship "User.addresses"; in SQLAlchemy 2.0, this reverse cascade will not take place.  Set cascade_backrefs to False in either the relationship() or backref() function for the 2.0 behavior; or to set globally for the whole Session, set the future=True flag (Background on SQLAlchemy 2.0 at: http://sqlalche.me/e/b8d9) (Background on this error at: http://sqlalche.me/e/14/s9r1)

yup as long as there's a place for people to go with the err message

oh also I see I put RST code into the error string in my review, that was a mistake on my part.


> ```
> 
> There are three parts of this that I'm not certain about, and I'd like to ask for advice/help getting this into a merge-ready state.
> 
>     1. Adding `code` to warning objects
> 
> 
> I looked at the stdlib's `warnings` to see how warnings are printed. [This line puts warnings in a small wrapper](https://github.com/python/cpython/blob/0983e01837714524fb164e784a8e96a2bc4bdf94/Lib/warnings.py#L394) and [then typically this is how that wrapper gets string-formatted](https://github.com/python/cpython/blob/0983e01837714524fb164e784a8e96a2bc4bdf94/Lib/warnings.py#L36-L37).
> 
> So all that we need to adjust is how `RemovedIn20Warning.__str__` behaves. That could be done by modifying the warning text, or by adjusting `__str__` directly. I opted for the latter, so `SADeprecationWarning` now has a modified `__str__` method.
> 
> I chose to do this by pulling `code` and `_code_str` from `SQLAlchemyError` into a mixin and then sharing that between both `SQLAlchemyError` and `SADeprecationWarning`. I wasn't sure if this would be preferable, vs copying the `code` and `_code_str` components into `SADeprecationWarning` without using the mixin, or some other approach.
> 
> I'm also not certain that this is all wanted or not?

the mixin looks great.  that's a fine approach.

we still need to have the corresponding entry in the errors.rst file.

> It's simpler to just modify the message if `code != None` in `sqlalchemy.util.deprecations`, but this puts plumbing in place for deprecations to have a `code` in general.
> 
> (Aside: If the warnings have custom `__str__` methods, maybe the special-case message text for `RemovedIn20Warning` should be handled there as well?)

if you can swing it and it makes things more consistent then sure

> 
>     1. Testing
> 
> 
> This passes existing tests today (at least, those which run under `tox -e py39-sqlite`, but it probably doesn't damage anything else).
> 
> At a minimum, some new test should cover the `SADeprecationWarning.code` behavior.

this is tested right now in test/orm/test_cascades.py -> test_m2o_backref_child_expunged and test_m2o_backref_child_expunged_nochange.  it checks the text of the message which could be expanded to include that the hyperlink is there.


> 
> I looked at the existing tests for the ORM cascade and for deprecations, and I was a bit overwhelmed. I would like to add a test that verifies that `http://sqlalche.me/e/{version}/s9r1` appears in the deprecation warnings now, but I don't know how to do so. Any guidance on how to test this change would be a big help.

it's tested e.g. here: https://github.com/sqlalchemy/sqlalchemy/blob/master/test/orm/test_cascade.py#L1384 .   i find these using grep (and of course I sort of know the general idea of where it will be which helps)

> 
>     1. Are the new docs good?

they look good so far, ill make more of a pass once the whole thing is in place


> 
> 
> The goal of these additions is primarily to make it clear that `backref.cascade_backrefs` exists, and you can read up on it if necessary. Was that accomplished? Would more narrative doc be helpful?

Let's get the entry in errors.rst and the linkage going, that is 90% of it, then we can tweak the narrative after the fact as the change here is more about the linkage infrastructure.


> As a user, I've only interacted with these features very minimally, so I'm not confident that I captured everything important or stated things well.


you're doing great this was a big change!

zzzeek commented:
ping me when i should look again at this, I will otherwise not remember

sirosen committed:
Update docs and HasDescriptionCode per review

Rename the mixin class and make a few changes to the docs. Notably,
put the '_error_s9r1' ReST ref in the errors doc. This requires a new
(small) section there to make sense.

sirosen committed:
Move custom 2.0 deprecation warning message

Rather than appending to the message via the deprecation utilities,
the RemovedIn20Warning class has a custom __str__ which appends the
desired info to its message.

sirosen committed:
Backref cascade warning test checks links

This is a deprecation test which already ensures that a warning si
emitted whenever a backref-cascade takes place.
Expand the text which it checks to include the two links which should
be appended to the message in this case.

sirosen commented:
@zzzeek, I've just finished a round of refinement here in response to your feedback. It turned out to be pretty easy to move the RemovedIn20Warning link into that warning's `__str__`, so that's included here too.

I hadn't realized that the link had to be to `errors.rst`, so I added a section there, which refers to the other relevant sections. (I originally added the rst ref to the ORM cascade doc, which I'm guessing, from what you've said, would not work?)

In https://github.com/sqlalchemy/sqlalchemy/pull/6250/commits/f23df2d147a3caf5d18e22f07909b4a7c8c4d95b , I expanded the existing deprecation test to check the warning links. I could add a new test just for this if that would be better.

zzzeek commented:
OK I am super backlogged at the moment as a whole lot of small regressions have been coming in.    this is on my list to look at thanks!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f23df2d147a3caf5d18e22f07909b4a7c8c4d95b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change f23df2d147a3caf5d18e22f07909b4a7c8c4d95b: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2775

sirosen commented:
I only just noticed that the CI build was failing on `tox -e pep8`, so I pushed a fixup after running `black`. (Missed a trailing comma.)

zzzeek commented:
hey thanks!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision a28e425117bfa2651225adc70e9e57f6c24d3dc1 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset a28e425117bfa2651225adc70e9e57f6c24d3dc1 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2775

sirosen committed:
Make HasDescriptionCode work under py2

On py2.7, this class needs to inherit from `object` in order to use
`super()`. Otherwise, it fails a check in `super()` that makes sure
that its first element is a `type`.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision dbcaeb54e31517fe88f6f8c515f1024002675f13 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset dbcaeb54e31517fe88f6f8c515f1024002675f13 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2775

sirosen commented:
It looks to me like CI is now passing. `HasDescriptionCode` needed to inherit from `object` in order to use `super(HasDescriptionCode, self)` under py2.

So I think we're finally there, and this can go through final review ---> merge. I hope! 😅

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

openstack-recheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2775

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2775 has been **merged**. Congratulations! :)

zzzeek commented:
thanks for your efforts on this as well as your great patience waiting a full month for this!

sirosen commented:
Of course! I'm just happy to have been able to help, even a little, with sqlalchemy.
